### PR TITLE
dev-8412 API Nested Dropdown

### PR DESCRIPTION
**Pull request by Matt-DiPaolo**
## DEV-8412

Added the URL option as a data source for Nested Dropdowns 

## Testing Steps

Scenario: Open a Default Editor and choose the Dashboard Viz and in Configuration, upload dev-8412-Nested-Dropdown.json (attached to ticket, [https://websupport.cdc.gov/browse/DEV-8412] )

Expected: There are 2 Visualization boxes. The one on top is empty and the bottom one has a table in it.

Scenario: Click on the Impot Data Tab and select Load from URL

1. Click the "Always load from URL" checkbox
2. Enter `https://nccd-cove-public-api.apps.ecpaas-dev.cdc.gov/od-public?$datakey=brfss_prevalence_cove_explore_by_topic`
3. Press the Load button

Expected: Data should populate on the right hand side under Data Preview

4. Click the Configure tab
5. Drag a Filter Dropdowns viz into the Visualization area
6. Click on the Gear icon for the Table
7. In Select a dataset box select the URL, vertical, no, and then click continue
8. Click the tools icon of the Filter Dropdowns viz

Expected: Filters Dropdown Viz will open with a Configure Dashboard Filter on the left and a blank screen on the right

9. Click the Filter Accordion
10. Click Add Filter
11. Click the Down arrow button
12. Select URL in the Filter Type dropdown
13. Select Nested Dropdown in the Filter Style dropdown
14. In the API Endpoint text box, add `https://nccd-cove-public-api.apps.ecpaas-dev.cdc.gov/od-public?$datakey=brfss_prevalence_cove_explore_by_topic&$select=distinct%20Class,Topic`
15. In Value Selector enter Class
16. In Subgroup Value Selector enter Topic
17. In Used By select table1727…
18. Click on the filter that was populated on the right

Expected: A nested dropdown list will appear with Class as the bold group titles and the Topic will be the options under the group titles.

Scenario: Click on the Back to Dashboard button and click the Dashboard Preview tab

Expected: There will be the Dashboard filter on top with a Table below with data already populated.

19. Click on the Filter
20. Select Binge Drinking under Alcohol Consumption title

Expected: The Table will be filtered to just show Binge Drinking data

## Self Review

- I have added testing steps for reviewers
- My changes generate same key error when the Group is populated in the dropdown and the subgroup hasn't populated yet. Once the subgroup is populated then there are no same key errors.

### CODE REVIEW

File: packages/core/components/NestedDropdown/nestedDropdownHelpers.ts [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
     if (tierOne.match(newRegex)) return [tierOne, tierTwo]
-    const newOptions = tierTwo.filter(option => String(option).match(newRegex))
+    const newOptions = tierTwo.filter(option => String(option.subgroupText).match(newRegex))
     return [tierOne, newOptions]
   })
 
   return filterOptions
 }
+
+export const organizeNestedFilterData = () => {}
```
Comment (joshlacey) [inline packages/core/components/NestedDropdown/nestedDropdownHelpers.ts:21]: ??

File: packages/dashboard/src/types/APIFilter.ts [hunk 2]
```diff
@@ -1 +1,6 @@
-export type APIFilter = Record<'apiEndpoint' | 'valueSelector' | 'textSelector', string>
+type subGroupValues = Partial<Record<'subgroupValueSelector' | 'subgroupTextSelector', string>>
+
+export type APIFilter = Record<'apiEndpoint' | 'valueSelector' | 'textSelector', string> &
+  subGroupValues & {
+    values: [string, Record<string, string | number>[]]
```
Comment (joshlacey) [inline packages/dashboard/src/types/APIFilter.ts:5]: Why are we storing values on this type? the values should be calculated at runtime only. This type is meant for the configuration that is saved, not runtime data

File: packages/map/src/components/UsaMap/components/UsaMap.County.tsx [hunk 3]
```diff
@@ -11,7 +11,16 @@ import ErrorBoundary from '@cdc/core/components/ErrorBoundary'
```
Comment (joshlacey) [inline packages/map/src/components/UsaMap/components/UsaMap.County.tsx:11]: did you mean for changes to this file to be in your PR?

File: packages/core/components/NestedDropdown/tests/nestedDropdownHelpers.test.ts [hunk 4]
```diff
@@ -21,32 +99,194 @@ describe('handleSearchTerm()', () => {
...
+          }
+        ]
+      ]
     ]
     expect(result2).toEqual(expectedResult2)
     const result3 = handleSearchTerm('tomato', optsMemo)
     const expectedResult3: OptionsMemo = [
-      ['Fruits', ['Tomato']],
-      ['Vegetables', ['Tomato']]
+      [
```
Comment (joshlacey) [inline packages/core/components/NestedDropdown/tests/nestedDropdownHelpers.test.ts:268]: it seems like a better runtime data structure here would be a tuple. the text indexes could be optional. and here you have an ordered list of your nested dropdown options.

[  [value, text], [ [subValue, subText], [subValue, subText] ]  ]

**Review by joshlacey — CHANGES_REQUESTED** [review]

File: packages/map/src/components/UsaMap/components/UsaMap.County.tsx (hunk 3 above)
Comment (Matt-DiPaolo) [reply packages/map/src/components/UsaMap/components/UsaMap.County.tsx:11]: I don't remember making these changes or why I would've done it. I know that you made some changes before I took over the branch. If this isn't part of  your changes then no. I will remove them.

File: packages/core/components/NestedDropdown/NestedDropdown.tsx [hunk 5]
```diff
@@ -60,8 +65,12 @@ const Options: React.FC<{
           aria-labelledby={label}
           className={isTierOneExpanded ? '' : 'hide'}
         >
-          {currentOptions.map((tierTwo, tierTwoIndex) => {
-            const regionID = label + tierTwo
+          {/* {currentOptions.map((tierTwo, tierTwoIndex) => { */}
```
Comment (joshlacey) [inline packages/core/components/NestedDropdown/NestedDropdown.tsx:68]: remove comment

File: packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx [hunk 6]
```diff
@@ -23,7 +24,7 @@ type FilterEditorProps = {
 const FilterEditor: React.FC<FilterEditorProps> = ({ filter, config, updateFilterProp }) => {
   const [columns, setColumns] = useState<string[]>([])
   const transform = new DataTransform()
-  const filterStyles = ['dropdown', 'multiselect', 'nested-dropdown']
+  const filterStyles = ['dropdown', 'multi-select', 'nested-dropdown']
```
Comment (joshlacey) [inline packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx:27]: Should be able to get this list from FILTER_STYLES

File: packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx [hunk 7]
```diff
@@ -116,231 +117,130 @@ const FilterEditor: React.FC<FilterEditorProps> = ({ filter, config, updateFilte
...
+        )}
+
+        <div className={isNestedDropDown ? 'border border-dark p-1 my-1' : ''}>
           <TextField
-            label='Option Text Selector:'
-            value={filter.apiFilter?.textSelector}
-            updateField={(_section, _subSection, _key, value) => updateAPIFilter('textSelector', value)}
+            label={`${isSubgroup ? 'Subgroup ' : ''}Value Selector:`}
+            value={filter?.apiFilter?.[valueSelector] || ''}
+            updateField={(_section, _subSection, _key, value) => updateAPIFilter(valueSelector, value)}
```
Comment (joshlacey) [inline packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx:160]: the first param is supposed to be the key that you want to set.

File: packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx [hunk 8]
```diff
@@ -116,231 +117,130 @@ const FilterEditor: React.FC<FilterEditorProps> = ({ filter, config, updateFilte
...
             }
           />
+
           <TextField
-            label='Option Value Selector:'
-            value={filter.apiFilter?.valueSelector}
-            updateField={(_section, _subSection, _key, value) => updateAPIFilter('valueSelector', value)}
+            label={`${isSubgroup ? 'Subgroup ' : ''}Display Text Selector:`}
+            value={filter?.apiFilter?.[textSelector] || ''}
+            updateField={(_section, _subSection, _key, value) => updateAPIFilter(textSelector, value)}
```
Comment (joshlacey) [inline packages/dashboard/src/components/DashboardFilters/DashboardFiltersEditor/components/FilterEditor.tsx:179]: again I think 'textSelector' is supposed to be a string

File: packages/dashboard/src/helpers/apiFilterHelpers.ts [hunk 9]
```diff
@@ -46,6 +47,31 @@ export const getFilterValues = (data: Array<Object>, apiFilter: APIFilter): Drop
   return data.map(v => ({ text: v[textSelector || valueSelector], value: v[valueSelector] }))
 }
 
+export const getNestedDropdownValues = (data: Array<Object>, apiFilter: APIFilter) => {
```
Comment (joshlacey) [inline packages/dashboard/src/helpers/apiFilterHelpers.ts:50]: test?

File: packages/core/helpers/sortByOrderedValues.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { SharedFilter } from '@cdc/dashboard/src/types/SharedFilter'
+
+export const sortByOrderedValues = (values: any[], filter: SharedFilter): void => {
```
Comment (joshlacey) [inline packages/core/helpers/sortByOrderedValues.ts:3]: where is this used?

**Review by joshlacey — CHANGES_REQUESTED** [review]

File: packages/core/components/Filters/helpers/getNestedOptions.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import { SubGrouping } from '../../../types/VizFilter'
+import { NestedOptions, ValueTextPair } from '../../NestedDropdown/nestedDropdownHelpers'
+
+type GetOptionsMemoParams = {
+  orderedValues?: string[]
+  values: string[]
+  subGrouping: SubGrouping
+}
+
+export const getNestedOptions = ({ orderedValues, values, subGrouping }: GetOptionsMemoParams): NestedOptions => {
```
Comment (joshlacey) [inline packages/core/components/Filters/helpers/getNestedOptions.ts:10]: test

File: packages/dashboard/src/components/DashboardFilters/DashboardFilters.tsx [hunk 12]
```diff
@@ -21,37 +24,54 @@ const DashboardFilters: React.FC<DashboardFilterProps> = ({
   }
 
   const updateField = (_section, _subsection, fieldName, value) => {
-    handleOnChange(fieldName, value)
+    handleOnChange(fieldName, value) // fieldName is the sharedFilterIndex
+  }
+
+  const getNestedDropdownOptions = (options?: DropdownOptions): NestedOptions => {
```
Comment (joshlacey) [inline packages/dashboard/src/components/DashboardFilters/DashboardFilters.tsx:30]: test

File: packages/dashboard/src/helpers/apiFilterHelpers.ts [hunk 13]
```diff
@@ -32,17 +33,39 @@ const getParentParams = (
   if (!(_parents || []).length) return null
 
   return _parents.flatMap(filter => {
-    const key = filter.queryParameter || filter.apiFilter.valueSelector || ''
-    const value = filter.queuedActive || filter.active || ''
-    if (Array.isArray(value)) {
-      return value.map(_value => ({ key, value: _value.toString() }))
+    if (filter.filterStyle === FILTER_STYLE.nestedDropdown) {
```
Comment (joshlacey) [inline packages/dashboard/src/helpers/apiFilterHelpers.ts:36]: test

File: packages/dashboard/src/helpers/apiFilterHelpers.ts [hunk 14]
```diff
@@ -32,17 +33,39 @@ const getParentParams = (
...
+      }
+      return [{ key, value: value.toString() }]
     }
-    return [{ key, value: value.toString() }]
   })
 }
 
 export const getFilterValues = (data: Array<Object>, apiFilter: APIFilter): DropdownOptions => {
-  const { textSelector, valueSelector } = apiFilter
+  const { textSelector, valueSelector, subgroupTextSelector, subgroupValueSelector } = apiFilter
```
Comment (joshlacey) [inline packages/dashboard/src/helpers/apiFilterHelpers.ts:57]: test

File: packages/dashboard/src/helpers/applyQueuedActive.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { FILTER_STYLE } from '../types/FilterStyles'
+import { SharedFilter } from '../types/SharedFilter'
+
+export const applyQueuedActive = (sharedFilter: SharedFilter) => {
```
Comment (joshlacey) [inline packages/dashboard/src/helpers/applyQueuedActive.ts:4]: test

**Review by joshlacey — CHANGES_REQUESTED** [review]

**Review by joshlacey — APPROVED** [review]

**Pull request merged**